controller/src/main/java/io/pravega/controller/store/stream/ZKStoreHelper.java

```diff
@@ -62,6 +64,21 @@ public ZKStoreHelper(final CuratorFramework cf, Executor executor) {
         return getChildren("/store");
     }
 
+    CompletableFuture<List<CuratorTransactionResult>> transaction(List<CuratorOp> opList) {
+        final CompletableFuture<List<CuratorTransactionResult>> result = new CompletableFuture<>();
+        try {
+            // We don't know the accurate path on which this transaction failed.
+            // See also ZOOKEEPER-3290
+            String unknownPath = "<unknown>";
```

Comment:
why is the path is unknown?

Reply:
In previous version of this pull request I try to do as you suggested here. Here are several concerns prevent it from implemented.

1. As you commented previous, if we composite paths before send transaction request, it might cost several cycles.  
2. Follow 1, due to the implementation of `callback`, we cannot lazily compute the third argument. Thus it seems noway to defer the computation unless we re-implement `callback` to support `Function<List, String>` or programming low level(should be avoided IMHO).
3. Semantically, the third parameter `callback` is named `path`, while a composition of paths is a concrete error message that definitely not a path.

What do you think? In fact we don't know the accurate path on which this transaction failed. If we want to provide detailed error message, a refactor/addition of  `callback` looks like required.

Reply:
I see what you are saying.. 
I would rather have the correct path as that is captured in the Exception that we throw and is useful for debugging when its logged. 

So since the transaction is used only in orderer, we could have your original mechanism and spend some cycles there.